# **Applying Advanced Transformations**
- Steven Phillips
- December 22, 2022

## **Part I - Clean the files and combine them into one final DataFrame.**
- Final DataFrame name will be df_superhero_merged

### Load and copy the datasets - import all libraries

In [431]:
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

#### Superhero Info

In [432]:
df1 = pd.read_csv('/Users/steve/Documents/GitHub/Applying-Advanced-Transformations-Core/Data/superhero_info - superhero_info.csv')
df1.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [433]:
df_superhero_info = df1.copy()
df_superhero_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


#### Superhero Powers

In [434]:
df2 = pd.read_csv('/Users/steve/Documents/GitHub/Applying-Advanced-Transformations-Core/Data/superhero_powers - superhero_powers.csv')
df2.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [435]:
df_superhero_powers = df2.copy()
df_superhero_powers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


### Clean the data sets

In [436]:
df_superhero_info.shape

(463, 8)

In [437]:
df_superhero_powers.shape

(667, 2)

#### Check for missing entries

In [438]:
df_superhero_info.isna().sum()

Hero|Publisher    0
Gender            0
Race              0
Alignment         0
Hair color        0
Eye color         0
Skin color        0
Measurements      0
dtype: int64

In [439]:
df_superhero_powers.isna().sum()

hero_names    0
Powers        0
dtype: int64

In [440]:
df_superhero_info.nunique()

Hero|Publisher    457
Gender              3
Race               52
Alignment           4
Hair color         29
Eye color          21
Skin color         14
Measurements      287
dtype: int64

In [441]:
df_superhero_powers.nunique()

hero_names    667
Powers        608
dtype: int64

In [442]:
df_superhero_info.duplicated().sum()

0

In [443]:
df_superhero_powers.duplicated().sum()

0

### Split the Hero/Publisher Column

In [444]:
df_superhero_info['Hero|Publisher'].head()

0            A-Bomb|Marvel Comics
1    Abe Sapien|Dark Horse Comics
2              Abin Sur|DC Comics
3       Abomination|Marvel Comics
4     Absorbing Man|Marvel Comics
Name: Hero|Publisher, dtype: object

In [445]:
## adding expand=True
df_superhero_info['Hero|Publisher'].str.split('|',expand=True)

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [446]:
## save the 2 new columns into the dataframe
df_superhero_info[['Hero_Name','Publisher']] = df_superhero_info['Hero|Publisher'].str.split('|',expand=True)
df_superhero_info.head(2)

Hero|Publisher Gender           Race Alignment Hair color  \
0          A-Bomb|Marvel Comics   Male          Human      good    No Hair   
1  Abe Sapien|Dark Horse Comics   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color                                  Measurements  \
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

    Hero_Name          Publisher  
0      A-Bomb      Marvel Comics  
1  Abe Sapien  Dark Horse Comics

In [447]:
## drop the original column 
df_superhero_info = df_superhero_info.drop(columns=['Hero|Publisher'])
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements      Hero_Name  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [448]:
df_superhero_info.nunique()

Gender            3
Race             52
Alignment         4
Hair color       29
Eye color        21
Skin color       14
Measurements    287
Hero_Name       455
Publisher         9
dtype: int64

### Separate the Measurements column

#### Single quotes must be replaced by double quotes

In [449]:
## examining a single value from the measurements col
measures = df_superhero_info.loc[0,"Measurements"]
print(type(measures))
measures

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [450]:
measures = measures.replace("'",'"')
measures

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [451]:
## now we can use json.loads
fixed_measures = json.loads(measures)
print(type(fixed_measures))
fixed_measures

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

#### Apply to the entire Measurements column

In [452]:
## use .str.replace to replace all single quotes
df_superhero_info['Measurements'] = df_superhero_info['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
df_superhero_info['Measurements'] = df_superhero_info['Measurements'].apply(json.loads)
df_superhero_info['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [453]:
## check a single value after transformation
test_coord = df_superhero_info.loc[0, 'Measurements']
print(type(test_coord))
test_coord

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [454]:
height_weight = df_superhero_info['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [455]:
# concat Height_Weight with original dataframe
df_superhero_info = pd.concat((df_superhero_info, height_weight), axis = 1)
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements      Hero_Name  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg  
2          DC Comics  185.0 cm   90.0 kg  
3      Marvel Comics  203.0 cm  441.0 kg  
4      Marvel Comics  193.0 cm  122.0 kg

In [456]:
df_superhero_info.rename(columns = {'Height':'Height(cm)', 'Weight':'Weight(kg)'}, inplace = True)

In [457]:
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements      Hero_Name  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height(cm) Weight(kg)  
0      Marvel Comics   203.0 cm   441.0 kg  
1  Dark Horse Comics   191.0 cm    65.0 kg  
2          DC Comics   185.0 cm    90.0 kg  
3      Marvel Comics   203.0 cm   441.0 kg  
4      Marvel Comics   193.0 cm   122.0 kg

#### Remove (cm) and (kg) from Height(cm) and Weight(kg)

In [458]:
## adding expand=True
df_superhero_info['Height(cm)'].str.split(' ',expand=True)

0   1
0    203.0  cm
1    191.0  cm
2    185.0  cm
3    203.0  cm
4    193.0  cm
..     ...  ..
458  183.0  cm
459  165.0  cm
460   66.0  cm
461  170.0  cm
462  185.0  cm

[463 rows x 2 columns]

In [459]:
## save the 2 new columns into the dataframe
df_superhero_info[['Height(cm)','(cm)']] = df_superhero_info['Height(cm)'].str.split(' ',expand=True)
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements      Hero_Name  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height(cm) Weight(kg) (cm)  
0      Marvel Comics      203.0   441.0 kg   cm  
1  Dark Horse Comics      191.0    65.0 kg   cm  
2          DC Comics      185.0    90.0 kg   cm  
3      Marvel Comics      203.0   441.0 kg   cm  
4      Marvel Comics      193.0   122.0 kg   cm

In [460]:
## save the 2 new columns into the dataframe
df_superhero_info[['Weight(kg)','(kg)']] = df_superhero_info['Weight(kg)'].str.split(' ',expand=True)
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements      Hero_Name  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height(cm) Weight(kg) (cm) (kg)  
0      Marvel Comics      203.0      441.0   cm   kg  
1  Dark Horse Comics      191.0       65.0   cm   kg  
2          DC Comics      185.0       90.0   cm   kg  
3      Marvel Comics      203.0      441.0   cm   kg  
4      Marvel Comics      193.0      122.0   cm   kg

In [461]:
df_superhero_info['(cm)'].value_counts()

cm    463
Name: (cm), dtype: int64

In [462]:
df_superhero_info['(kg)'].value_counts()

kg    463
Name: (kg), dtype: int64

In [463]:
# drop (cm) and (kg) columns
df_superhero_info = df_superhero_info.drop(columns = ['(cm)', '(kg)'])
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements      Hero_Name  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height(cm) Weight(kg)  
0      Marvel Comics      203.0      441.0  
1  Dark Horse Comics      191.0       65.0  
2          DC Comics      185.0       90.0  
3      Marvel Comics      203.0      441.0  
4      Marvel Comics      193.0      122.0

In [464]:
# drop original Measurments column
df_superhero_info = df_superhero_info.drop(columns = ['Measurements'])
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

       Hero_Name          Publisher Height(cm) Weight(kg)  
0         A-Bomb      Marvel Comics      203.0      441.0  
1     Abe Sapien  Dark Horse Comics      191.0       65.0  
2       Abin Sur          DC Comics      185.0       90.0  
3    Abomination      Marvel Comics      203.0      441.0  
4  Absorbing Man      Marvel Comics      193.0      122.0

In [465]:
# rename Hero(name) in both data frames
df_superhero_info.rename(columns = {'Hero_Name':'Hero(name)'}, inplace = True)
df_superhero_info.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

      Hero(name)          Publisher Height(cm) Weight(kg)  
0         A-Bomb      Marvel Comics      203.0      441.0  
1     Abe Sapien  Dark Horse Comics      191.0       65.0  
2       Abin Sur          DC Comics      185.0       90.0  
3    Abomination      Marvel Comics      203.0      441.0  
4  Absorbing Man      Marvel Comics      193.0      122.0

In [466]:
# rename Hero(name) in both data frames
df_superhero_powers.rename(columns = {'hero_names':'Hero(name)'}, inplace = True)
df_superhero_powers.head()

Hero(name)                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

#### Change Height(cm) and Weight(cm) to numeric

In [467]:
df_superhero_info['Height(cm)'] = df_superhero_info['Height(cm)'].astype("float")
df_superhero_info['Weight(kg)'] = df_superhero_info['Weight(kg)'].astype("float")

In [468]:
df_superhero_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Gender      463 non-null    object 
 1   Race        463 non-null    object 
 2   Alignment   463 non-null    object 
 3   Hair color  463 non-null    object 
 4   Eye color   463 non-null    object 
 5   Skin color  463 non-null    object 
 6   Hero(name)  463 non-null    object 
 7   Publisher   463 non-null    object 
 8   Height(cm)  463 non-null    float64
 9   Weight(kg)  463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


### One-hot-encode the superpowers

In [469]:
df_superhero_powers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Hero(name)  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


In [470]:
df_superhero_powers['Powers'].value_counts()

Intelligence                                                                                                                                                                                                                                                         8
Durability,Super Strength                                                                                                                                                                                                                                            5
Agility,Stealth,Marksmanship,Weapons Master,Stamina                                                                                                                                                                                                                  4
Marksmanship                                                                                                                                                                                                       

In [471]:
## showing the lists are really strings
df_superhero_powers.loc[2,'Powers']

'Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner'

In [472]:
df_superhero_powers.shape

(667, 2)

In [473]:
# add a column that converts column of strings to column of lists
list_adjustment = []
for i in range(0,667):
    temporary = df_superhero_powers.loc[i, 'Powers'].split(",")
    list_adjustment.append(temporary)
df_superhero_powers['Powers_list'] = list_adjustment

In [474]:
# confirm column of powers as lists
df_superhero_powers['Powers_list'].head()

0      [Agility, Super Strength, Stamina, Super Speed]
1    [Accelerated Healing, Durability, Longevity, S...
2    [Agility, Accelerated Healing, Cold Resistance...
3                                 [Lantern Power Ring]
4    [Accelerated Healing, Intelligence, Super Stre...
Name: Powers_list, dtype: object

In [475]:
## exploding the column of lists
exploded = df_superhero_powers.explode('Powers_list')
exploded[['Hero(name)','Powers','Powers_list']].head(10)

Hero(name)                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

           Powers_list  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing  
1           Durability  
1            Longevity  
1       Super Strength  
1              Stamina  
1           Camouflage

In [476]:
## saving the unique values from the exploded column
cols_to_make = exploded['Powers_list'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [477]:
# for loop to add a column for each power and insert True/False for each superhero
for col in cols_to_make:
    df_superhero_powers[col] = df_superhero_powers['Powers'].str.contains(col)
df_superhero_powers.head()

C:\Users\steve\AppData\Local\Temp\ipykernel_64964\3387347489.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_superhero_powers[col] = df_superhero_powers['Powers'].str.contains(col)
C:\Users\steve\AppData\Local\Temp\ipykernel_64964\3387347489.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df_superhero_powers[col] = df_superhero_powers['Powers'].str.contains(col)
C:\Users\steve\AppData\Local\Temp\ipykernel_64964\3387347489.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling 

Hero(name)                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                         Powers_list  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [478]:
# drop Powers_list and original Powers columns
df_superhero_powers = df_superhero_powers.drop(columns=['Powers','Powers_list'])


### Join the two data frames to df_superhero_merged

In [479]:
## merge our 2 dataframes
df_superhero_merged = pd.merge(df_superhero_info, df_superhero_powers, on='Hero(name)')
df_superhero_merged.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

      Hero(name)          Publisher  Height(cm)  Weight(kg)  ...  \
0         A-Bomb      Marvel Comics       203.0       441.0  ...   
1     Abe Sapien  Dark Horse Comics       191.0        65.0  ...   
2       Abin Sur          DC Comics       185.0        90.0  ...   
3    Abomination      Marvel Comics       203.0       441.0  ...   
4  Absorbing Man      Marvel Comics       193.0       122.0  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 177 columns]

## **Part II - Use your combined DataFrame to answer the following questions:**
- 1. Compare the average weight of super powers who have Super Speed to those who do not.
- 2. What is the average height of heroes for each publisher?

### 1. Compare the average weight of super heroes who have Super Speed to those who do not.

In [480]:
df_superhero_merged['Super Speed']

0      False
1      False
2      False
3       True
4      False
       ...  
458    False
459    False
460     True
461    False
462     True
Name: Super Speed, Length: 463, dtype: bool

In [481]:
# super speed filter
ss_filter = df_superhero_merged['Super Speed'] == True
ss_filter

0      False
1      False
2      False
3       True
4      False
       ...  
458    False
459    False
460     True
461    False
462     True
Name: Super Speed, Length: 463, dtype: bool

In [482]:
# average weight of all superheros
df_superhero_merged['Weight(kg)'].mean().round(2)

113.59

In [483]:
# average weight of Super Speed superheros
df_superhero_merged.loc[ss_filter, 'Weight(kg)'].mean().round(2)

129.4

In [484]:
# average weight of superheros without Super Speed
df_superhero_merged.loc[~ss_filter, 'Weight(kg)'].mean().round(2)

101.77

The superheros with Super Speed have an average weight of 129.4 kg.  The superheros without Super Speed have an average weight of 101.77 kg.

In [485]:
# use groupby to find the same results
df_superhero_merged.groupby(['Super Speed'])['Weight(kg)'].mean().round(2)

Super Speed
False    101.77
True     129.40
Name: Weight(kg), dtype: float64

### 2. What is the average height of heroes for each publisher?

In [486]:
# output a list of average height of heroes for each publisher using groupby
df_superhero_merged.groupby(['Publisher'])['Height(cm)'].mean().round(2)

Publisher
DC Comics            181.92
Dark Horse Comics    176.91
George Lucas         159.60
Image Comics         211.00
Marvel Comics        191.55
Shueisha             171.50
Star Trek            181.50
Team Epic TV         180.75
Unknown              178.00
Name: Height(cm), dtype: float64